File: BoardSight/ios/BoardSight/CVModule.swift
import AVFoundation
import CoreImage
import Foundation
import React

/// CVModule — native iOS board-detection and move-detection module.
///
/// Architecture (MVP — Option A: grid + frame diff):
///   1. AVCaptureSession delivers CMSampleBuffers at ~15fps.
///   2. BoardDetector finds the 4 board corners via Canny + Hough lines,
///      emits BoardObservation when confidence > threshold.
///   3. After calibration, OccupancyTracker hashes each of 64 squares per frame
///      and emits MoveCandidate when a diff is stable across N frames.
///
/// TODO post-MVP: replace OccupancyTracker with a CoreML piece classifier.

@objc(CVModuleNative)
class CVModuleNative: RCTEventEmitter {

  // MARK: - Session state

  private var captureSession: AVCaptureSession?
  private var videoOutput: AVCaptureVideoDataOutput?
  private var sessionQueue = DispatchQueue(label: "com.boardsight.cv.session")
  private var processingQueue = DispatchQueue(label: "com.boardsight.cv.processing",
                                               qos: .userInitiated)

  // MARK: - Config

  private var confidenceThreshold: Double = 0.85
  private var targetFps: Int = 15
  private var enablePositionObs: Bool = false

  // MARK: - Detection state

  private var isTracking: Bool = false
  private var calibrationData: NSDictionary?

  /// Consecutive frame buffer for temporal smoothing (move detection)
  private var frameDiffHistory: [[String: String]] = []
  private let smoothingFrames = 3 // require N identical diffs

  // MARK: - RCTEventEmitter overrides

  override static func requiresMainQueueSetup() -> Bool { false }

  override func supportedEvents() -> [String]! {
    ["onBoardObservation", "onMoveCandidate", "onPositionObservation"]
  }

  // MARK: - JS-exposed commands

  @objc func startSession(_ config: NSDictionary) {
    confidenceThreshold = (config["confidenceThreshold"] as? Double) ?? 0.85
    targetFps = (config["targetFps"] as? Int) ?? 15
    enablePositionObs = (config["enablePositionObs"] as? Bool) ?? false

    sessionQueue.async { [weak self] in
      self?.setupCaptureSession()
    }
  }

  @objc func stopSession() {
    sessionQueue.async { [weak self] in
      self?.captureSession?.stopRunning()
      self?.captureSession = nil
      self?.videoOutput = nil
      self?.isTracking = false
    }
  }

  @objc func pauseTracking(_ paused: Bool) {
    isTracking = !paused
  }

  @objc func setCalibration(_ calib: NSDictionary) {
    calibrationData = calib
    isTracking = true
    frameDiffHistory.removeAll()
  }

  @objc func requestKeyFrame() {
    // Debug only — saves annotated frame to Photos
    guard _isDebugAssertConfiguration() else { return }
    // TODO: capture next frame and save to PHPhotoLibrary
  }

  // MARK: - AVCaptureSession setup

  private func setupCaptureSession() {
    let session = AVCaptureSession()
    session.sessionPreset = .hd1280x720

    guard
      let device = AVCaptureDevice.default(.builtInWideAngleCamera,
                                           for: .video,
                                           position: .back),
      let input = try? AVCaptureDeviceInput(device: device),
      session.canAddInput(input)
    else {
      NSLog("[CVModule] Failed to set up capture device")
      return
    }

    // Configure target frame rate
    try? device.lockForConfiguration()
    let desiredFps = CMTimeMake(value: 1, timescale: Int32(targetFps))
    device.activeVideoMinFrameDuration = desiredFps
    device.activeVideoMaxFrameDuration = desiredFps
    device.unlockForConfiguration()

    session.addInput(input)

    let output = AVCaptureVideoDataOutput()
    output.videoSettings = [kCVPixelBufferPixelFormatTypeKey as String:
                              kCVPixelFormatType_32BGRA]
    output.alwaysDiscardsLateVideoFrames = true
    output.setSampleBufferDelegate(self, queue: processingQueue)

    guard session.canAddOutput(output) else { return }
    session.addOutput(output)

    captureSession = session
    videoOutput = output
    session.startRunning()
    NSLog("[CVModule] Capture session started at %dfps", targetFps)
  }

  // MARK: - Board detection helpers (stubs — implement with Vision / OpenCV)

  /// Returns confidence 0–1 that a chessboard is visible, plus approximate corners.
  /// TODO: Use Vision framework's VNDetectRectanglesRequest + Hough line refinement.
  private func detectBoard(in pixelBuffer: CVPixelBuffer) -> (confidence: Double,
                                                               corners: [[String: Double]],
                                                               lightingWarning: Bool) {
    // STUB — replace with real CV
    return (confidence: 0.0, corners: [], lightingWarning: false)
  }

  /// Returns a per-square occupancy hash map [square → hash].
  /// TODO: After calibration, warp each square region and hash pixel values.
  private func hashOccupancy(in pixelBuffer: CVPixelBuffer) -> [String: String] {
    // STUB — replace with real CV
    return [:]
  }
}

// MARK: - AVCaptureVideoDataOutputSampleBufferDelegate

extension CVModuleNative: AVCaptureVideoDataOutputSampleBufferDelegate {

  func captureOutput(_ output: AVCaptureOutput,
                     didOutput sampleBuffer: CMSampleBuffer,
                     from connection: AVCaptureConnection) {
    guard let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else { return }
    let timestamp = Int64(Date().timeIntervalSince1970 * 1000)

    if calibrationData == nil {
      // Phase 1 — looking for board
      let (confidence, corners, lightingWarning) = detectBoard(in: pixelBuffer)
      guard confidence > 0 else { return }

      let event: NSDictionary = [
        "corners": corners,
        "confidence": confidence,
        "lightingWarning": lightingWarning,
        "timestamp": timestamp,
      ]
      sendEvent("onBoardObservation", body: event)

    } else if isTracking {
      // Phase 2 — tracking moves
      let occupancy = hashOccupancy(in: pixelBuffer)
      guard !occupancy.isEmpty else { return }

      // Temporal smoothing: accumulate frame diffs
      frameDiffHistory.append(occupancy)
      if frameDiffHistory.count > smoothingFrames {
        frameDiffHistory.removeFirst()
      }

      // TODO: Compare consecutive frames, detect stable 2-square diff → MoveCandidate
      // For now the detection is stubbed — real implementation in M1 CV work
    }
  }
}

File: BoardSight/android/app/src/main/java/com/boardsight/CVModule.kt
package com.boardsight

import android.Manifest
import android.content.pm.PackageManager
import android.graphics.ImageFormat
import android.hardware.camera2.*
import android.media.ImageReader
import android.os.Handler
import android.os.HandlerThread
import android.util.Log
import androidx.core.content.ContextCompat
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule

/**
 * CVModule — Android native CV module for board detection and move tracking.
 *
 * Architecture (MVP — Option A: grid + frame diff):
 *   1. Camera2 API delivers frames at ~15fps via ImageReader.
 *   2. BoardDetector finds board corners (Canny + Hough lines via OpenCV JNI — TODO).
 *   3. OccupancyTracker hashes 64 squares per frame; stable diff → MoveCandidate.
 *
 * TODO post-MVP: replace OccupancyTracker with TFLite piece classifier.
 */
class CVModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    companion object {
        private const val TAG = "CVModule"
        private const val TARGET_WIDTH = 1280
        private const val TARGET_HEIGHT = 720
        private const val DEFAULT_FPS = 15
        private const val SMOOTHING_FRAMES = 3
    }

    // Camera2 handles
    private var cameraManager: CameraManager? = null
    private var cameraDevice: CameraDevice? = null
    private var captureSession: CameraCaptureSession? = null
    private var imageReader: ImageReader? = null
    private var backgroundThread: HandlerThread? = null
    private var backgroundHandler: Handler? = null

    // Config
    private var confidenceThreshold: Double = 0.85
    private var targetFps: Int = DEFAULT_FPS
    private var enablePositionObs: Boolean = false

    // Detection state
    private var isTracking: Boolean = false
    private var calibrationData: ReadableMap? = null
    private val frameDiffHistory = mutableListOf<Map<String, String>>()

    override fun getName() = "CVModuleNative"

    // -------------------------------------------------------------------------
    // JS-exposed methods
    // -------------------------------------------------------------------------

    @ReactMethod
    fun startSession(config: ReadableMap) {
        confidenceThreshold = if (config.hasKey("confidenceThreshold"))
            config.getDouble("confidenceThreshold") else 0.85
        targetFps = if (config.hasKey("targetFps")) config.getInt("targetFps") else DEFAULT_FPS
        enablePositionObs = config.hasKey("enablePositionObs") && config.getBoolean("enablePositionObs")

        startBackgroundThread()
        openCamera()
    }

    @ReactMethod
    fun stopSession() {
        captureSession?.close()
        cameraDevice?.close()
        imageReader?.close()
        stopBackgroundThread()
        captureSession = null
        cameraDevice = null
        imageReader = null
        isTracking = false
    }

    @ReactMethod
    fun pauseTracking(paused: Boolean) {
        isTracking = !paused
    }

    @ReactMethod
    fun setCalibration(calib: ReadableMap) {
        calibrationData = calib
        isTracking = true
        frameDiffHistory.clear()
    }

    @ReactMethod
    fun requestKeyFrame() {
        // Debug only — saves annotated frame
        if (BuildConfig.DEBUG) {
            Log.d(TAG, "requestKeyFrame — TODO: save annotated frame to gallery")
        }
    }

    // -------------------------------------------------------------------------
    // Camera2 setup
    // -------------------------------------------------------------------------

    private fun openCamera() {
        val ctx = reactApplicationContext
        if (ContextCompat.checkSelfPermission(ctx, Manifest.permission.CAMERA)
            != PackageManager.PERMISSION_GRANTED) {
            Log.e(TAG, "Camera permission not granted")
            return
        }

        val manager = ctx.getSystemService(CameraManager::class.java)
        cameraManager = manager

        // Find back-facing camera
        val cameraId = manager.cameraIdList.firstOrNull { id ->
            manager.getCameraCharacteristics(id)
                .get(CameraCharacteristics.LENS_FACING) == CameraCharacteristics.LENS_FACING_BACK
        } ?: return

        val reader = ImageReader.newInstance(TARGET_WIDTH, TARGET_HEIGHT,
            ImageFormat.YUV_420_888, 2)
        reader.setOnImageAvailableListener({ r ->
            val image = r.acquireLatestImage() ?: return@setOnImageAvailableListener
            try {
                processFrame(image)
            } finally {
                image.close()
            }
        }, backgroundHandler)
        imageReader = reader

        manager.openCamera(cameraId, object : CameraDevice.StateCallback() {
            override fun onOpened(camera: CameraDevice) {
                cameraDevice = camera
                createCaptureSession(camera, reader)
            }
            override fun onDisconnected(camera: CameraDevice) { camera.close() }
            override fun onError(camera: CameraDevice, error: Int) {
                Log.e(TAG, "Camera error: $error")
                camera.close()
            }
        }, backgroundHandler)
    }

    private fun createCaptureSession(camera: CameraDevice, reader: ImageReader) {
        val surfaces = listOf(reader.surface)
        camera.createCaptureSession(surfaces, object : CameraCaptureSession.StateCallback() {
            override fun onConfigured(session: CameraCaptureSession) {
                captureSession = session
                val request = camera.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW).apply {
                    addTarget(reader.surface)
                    set(CaptureRequest.CONTROL_AE_TARGET_FPS_RANGE,
                        android.util.Range(targetFps, targetFps))
                }.build()
                session.setRepeatingRequest(request, null, backgroundHandler)
                Log.d(TAG, "Capture session started at ${targetFps}fps")
            }
            override fun onConfigureFailed(session: CameraCaptureSession) {
                Log.e(TAG, "Capture session configuration failed")
            }
        }, backgroundHandler)
    }

    // -------------------------------------------------------------------------
    // Frame processing (stubs — real CV in M1)
    // -------------------------------------------------------------------------

    private fun processFrame(image: android.media.Image) {
        val timestamp = System.currentTimeMillis()

        if (calibrationData == null) {
            // Phase 1: board detection stub
            // TODO: Convert YUV → grayscale, run Canny + Hough, detect 9x9 grid
            val confidence = 0.0 // placeholder
            if (confidence > confidenceThreshold) {
                val event = Arguments.createMap().apply {
                    putArray("corners", Arguments.createArray())
                    putDouble("confidence", confidence)
                    putBoolean("lightingWarning", false)
                    putDouble("timestamp", timestamp.toDouble())
                }
                sendEvent("onBoardObservation", event)
            }
        } else if (isTracking) {
            // Phase 2: occupancy hashing stub
            // TODO: Warp each of 64 squares via homography, hash pixel values
            // TODO: Compare consecutive hashes, emit MoveCandidate on stable diff
        }
    }

    // -------------------------------------------------------------------------
    // Event emission
    // -------------------------------------------------------------------------

    private fun sendEvent(eventName: String, params: WritableMap) {
        reactContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(eventName, params)
    }

    // -------------------------------------------------------------------------
    // Background thread
    // -------------------------------------------------------------------------

    private fun startBackgroundThread() {
        backgroundThread = HandlerThread("CVBackgroundThread").also {
            it.start()
            backgroundHandler = Handler(it.looper)
        }
    }

    private fun stopBackgroundThread() {
        backgroundThread?.quitSafely()
        backgroundThread?.join()
        backgroundThread = null
        backgroundHandler = null
    }
}
